include/bitcoin/server/web/socket.hpp

```diff
@@ -0,0 +1,169 @@
+namespace server {
+
+class server_node;
+
+class BCS_API socket
+  : public bc::protocol::zmq::worker
+{
+public:
+    // These "dumb" pointers are managed internally by mongoose and
+    // exposed to us via their API.
+    typedef struct mg_connection* connection_ptr;
+
+    // Tracks websocket queries via the query_work_map.  Used for
+    // matching websocket client requests to zmq query responses.
+    struct query_work_item
+    {
+        // Constructor provided for in-place construction.
+        query_work_item(uint32_t id, uint32_t correlation_id,
+            connection_ptr connection, const std::string& command,
+            const std::string& arguments)
+        :   id(id),
+            correlation_id(correlation_id),
+            command(command),
+            arguments(arguments),
+            connection(connection)
+        {
+        }
+
+        uint32_t id;
+        uint32_t correlation_id;
+        std::string command;
+        std::string arguments;
+        connection_ptr connection;
+    };
+
+    typedef std::function<void(bc::protocol::zmq::message&, const std::string&,
+        const std::string&, const uint32_t)> encode_handler;
+
+    typedef std::function<void(const data_chunk&, const uint32_t,
+        connection_ptr)> decode_handler;
+
+    // Handles translation of incoming JSON to zmq protocol methods and
+    // converting the result back to JSON for web clients.
+    struct handlers
+    {
+        std::string command;
+        encode_handler encode;
+        decode_handler decode;
+    };
+
+    typedef std::unordered_map<uint32_t, query_work_item> query_work_map;
+    typedef std::unordered_map<connection_ptr, query_work_map>
+        connection_work_map;
+    typedef std::unordered_map<uint32_t, std::pair<connection_ptr, uint32_t>>
+        query_correlation_map;
+    typedef std::unordered_map<std::string, handlers> handler_map;
+
+    /// Construct a socket class.
+    socket(bc::protocol::zmq::authenticator& authenticator, server_node& node,
+        bool secure, const std::string& domain);
+
+    /// Start the service.
+    bool start() override;
+
+    size_t connection_count() const;
+    void add_connection(connection_ptr connection);
+    void remove_connection(connection_ptr connection);
+    void notify_query_work(connection_ptr connection,
+        const std::string& command, const uint32_t sequence,
+        const std::string& arguments);
+
+protected:
+    typedef struct mg_mgr connection_manager;
+
+    // Initialize the websocket event loop and start a thread to poll events.
+    virtual bool start_websocket_handler();
+
+    // Terminate the websocket event loop.
+    virtual bool stop_websocket_handler();
+
+    virtual void handle_websockets();
+
+    virtual const config::endpoint& retrieve_zeromq_endpoint() const = 0;
+    virtual const config::endpoint& retrieve_websocket_endpoint() const = 0;
+    virtual const std::shared_ptr<bc::protocol::zmq::socket> get_service() const;
```

Comment:
The naming convention for getters is simply `zeromq_endpoint()` (no `retrieve_`) and `service()` (no `get_`).

Reply:
Ah right, I like that.